### DISCUSSION TITLE
Is there a way to do stable random search?

### DISCUSSION THREAD

**Discussion by lanyusan** [milvus-io/milvus#23866] [Q&A and General discussion] [answered]
My use case is that I wish to randomly present a feed to each refresh by user for items in a certain collection, for example photos.

Yet when user scroll down, this random generated feed should be stable and have no duplicates or few duplicates.

Each refresh could generate a random seed, then this seed can be used for this search. 

I am not sure if it is an appropriate application of vector search. Any insight would be much appreciated.

Thanks.

**Comment by xiaofan-luan** [comment]
> My use case is that I wish to randomly present a feed to each refresh by user for items in a certain collection, for example photos.
> 
> Yet when user scroll down, this random generated feed should be stable and have no duplicates or few duplicates.
> 
> Each refresh could generate a random seed, then this seed can be used for this search.
> 
> I am not sure if it is an appropriate application of vector search. Any insight would be much appreciated.
> 
> Thanks.

What if you just do the pagination? you can page from top 1 to top 65536. what you just need to be do is to randomly pick an offset? doesn't that sounds good to you?

**Reply by lanyusan** [reply]
This approach isn't what I wanted, because in each page the items are correlated by certain algorithm. I wish each page is completely random.

The reason is pretty simple. I don't want to predict or learn what user may like. I wish to present users with surprises through totally random feed. They can pick what they haven't seen before, open it, then a similarity search can give the user more similar ones.

**Reply by xiaofan-luan** [reply]
> This approach isn't what I wanted, because in each page the items are correlated by certain algorithm. I wish each page is completely random.
> 
> The reason is pretty simple. I don't want to predict or learn what user may like. I wish to present users with surprises through totally random feed. They can pick what they haven't seen before, open it, then a similarity search can give the user more similar ones.

So when you say random, is it like pure random? Then it's not actually ANN but randomly pick some vectors?

**Reply by yhmo** [reply]
> This approach isn't what I wanted, because in each page the items are correlated by certain algorithm. I wish each page is completely random.
> 
> The reason is pretty simple. I don't want to predict or learn what user may like. I wish to present users with surprises through totally random feed. They can pick what they haven't seen before, open it, then a similarity search can give the user more similar ones.

Correct me if my understanding is incorrect:
- Assume you have 1 million photos
- You convert these photos to 1 million vectors, let's say the dimension is 512
- Your APP connects to a Milvus server, the 1 millions vectors are stored in Milvus
- Your user refreshes the page in the APP, the APP generates a random vector(dim=512) by a random seed(might be a timestamp)
- The APP call search() to find top20 similar vectors from the Milvus server, and return the corresponding photos
- The APP presents the photos on the page. Each time user refreshes the page, he will see different photos

**Reply by lanyusan** [reply]
Yes. I don't know if there is any algorithm to generate random distance by a seed. If so then problem solved. Other wise, I am thinking of pre-building  hundreds of random feeds in advance, update and regenerate them periodically, then use similarity search for individual items of interests for users.

**Reply by lanyusan** [reply]
```
Correct me if my understanding is incorrect:

Assume you have 1 million photos
You convert these photos to 1 million vectors, let's say the dimension is 512
Your APP connects to a Milvus server
Your user refreshes the page in the APP, the APP generates a random vector(dim=512) by a random seed(might be a timestamp)
The APP call search() to find top20 similar vectors from the Milvus server, and return the corresponding photos
The APP presents the photos on the page. Each time user refreshes the page, he will see different photos
```

This isn't what i wanted either, because this random vector could be interpreted as an image or something with certain features, then search results are all ranked by similarity. 

Ideally the distance calculation algorithm should be random.

**Reply by yhmo** [reply]
In my opinion, no need to calculate a "random distance". You can use "random ID" instead of "random distance".
- Assume there are 1 million photos
- Convert the photos to 1 million vectors
- Insert the vectors into a Milvus server, each vector has an ID (int64 type), from 1 to 1000000
- Your APP connect to the Milvus server
- Your user refreshes the page in the APP, the APP generates an ID base on a random seed
- Pick a vector by this ID from the Milvus server by query() interface
- Use the vector to search in Milvus, find out the topk similar vectors
- Show similar photos on the page

It is easy to generate a random integer ID, rather than generating a random distance, right?

**Reply by lanyusan** [reply]
> In my opinion, no need to calculate a "random distance". You can use "random ID" instead of "random distance".
> 
> * Assume there are 1 million photos
> * Convert the photos to 1 million vectors
> * Insert the vectors into a Milvus server, each vector has an ID (int64 type), from 1 to 1000000
> * Your APP connect to the Milvus server
> * Your user refreshes the page in the APP, the APP generates an ID base on a random seed
> * Pick a vector by this ID from the Milvus server by query() interface
> * Use the vector to search in Milvus, find out the topk similar vectors
> * Show similar photos on the page
> 
> It is easy to generate a random integer ID, rather than generating a random distance, right?

You mean use a random int as a one dimension vector, and then generate a random int to calculate distance so the result is random?
It sounds to be what I needed.

Thanks.

**Reply by yhmo** [reply, marked as answer]
> > In my opinion, no need to calculate a "random distance". You can use "random ID" instead of "random distance".
> > 
> > * Assume there are 1 million photos
> > * Convert the photos to 1 million vectors
> > * Insert the vectors into a Milvus server, each vector has an ID (int64 type), from 1 to 1000000
> > * Your APP connect to the Milvus server
> > * Your user refreshes the page in the APP, the APP generates an ID base on a random seed
> > * Pick a vector by this ID from the Milvus server by query() interface
> > * Use the vector to search in Milvus, find out the topk similar vectors
> > * Show similar photos on the page
> > 
> > It is easy to generate a random integer ID, rather than generating a random distance, right?
> 
> You mean use a random int as a one dimension vector, and then generate a random int to calculate distance so the result is random? It sounds to be what I needed.
> 
> Thanks.

A random int to pick a vector to calculate distance.

```
import random

from pymilvus import (
    connections,
    FieldSchema, CollectionSchema, DataType,
    Collection, SearchFuture, SearchResult,
    utility,
)


if __name__ == '__main__':
    connections.connect(host='127.0.0.1', port="19530")

    utility.get_server_version()
    id_field = "id"
    vector_field = "vector"
    dim = 128
    collection_name = "demo"

    if utility.has_collection(collection_name):
        utility.drop_collection(collection_name)

    # create a collection with two fields: "id" and "vector"
    field1 = FieldSchema(name=id_field, dtype=DataType.INT64, is_primary=True, auto_id=False)
    field2 = FieldSchema(name=vector_field, dtype=DataType.FLOAT_VECTOR, dim=dim)
    schema = CollectionSchema(fields=[field1, field2])
    collection = Collection(name=collection_name, schema=schema, shards_num=1)
    print("create collection")
    index_param = {"index_type": "HNSW", "metric_type": "L2", "params": {"M": 48, "efConstruction": 500}}
    collection.create_index(vector_field, index_param)
    collection.load()

    # insert 1000 rows into the collection, each row has an ID and a vector, the ID is from 0 to 999
    count = 1000
    data = [
        [k for k in range(count)],
        [[random.random() for _ in range(dim)] for _ in range(count)],
    ]
    collection.insert(data)
    print("insert done")

    collection = Collection(collection_name)
    # generate a random ID to pick a vector
    id = random.randint(0, 999)
    expr = "id==" + str(id)
    print("query vector by", expr)
    res = collection.query(expr=expr, output_fields=[vector_field])
    if len(res) == 0:
        raise Exception("Failed to retrieve vector")

    # use the random vector to search
    print("search random vector")
    vector = res[0][vector_field]
    search_param = {
        "data": [vector],
        "anns_field": vector_field,
        "param": {"metric_type": "L2", "params": {"ef": 16}},
        "limit": 10,
    }
    results = collection.search(**search_param)
    for hits in results:
        for hit in hits:
            print(hit)

```

**Reply by lanyusan** [reply]
Thanks
